pkg/kubectl/cmd/annotation.go

```diff
@@ -0,0 +1,211 @@
+		Example: annotation_example,
+		Run: func(cmd *cobra.Command, args []string) {
+			err := RunAnnotation(f, out, cmd, args)
+			cmdutil.CheckErr(err)
+		},
+	}
+	cmdutil.AddPrinterFlags(cmd)
+	cmd.Flags().Bool("overwrite", false, "If true, allow annotations to be overwritten, otherwise reject annotation updates that overwrite existing annotations.")
+	cmd.Flags().Bool("all", false, "select all resources in the namespace of the specified resource types")
+	cmd.Flags().String("resource-version", "", "If non-empty, the annotation update will only succeed if this is the current resource-version for the object. Only valid when specifying a single resource.")
+	return cmd
+}
+
+// validateNoAnnotationOverwrites validates that when overwrite is false, to-be-updated annotations don't exist in the object annotation map (yet)
+func validateNoAnnotationOverwrites(meta *api.ObjectMeta, annotations map[string]string) error {
+	for key := range annotations {
+		if value, found := meta.Annotations[key]; found {
+			return fmt.Errorf("'%s' already has a value (%s), and --overwrite is false", key, value)
+		}
+	}
+	return nil
+}
+
+// parseAnnotations parses the annotaion arg and returns annotation maps to update and annotaions to remove
+func parseAnnotations(spec []string) (annotations map[string]string, remove []string, err error) {
+	annotations = map[string]string{}
+	for _, annotationSpec := range spec {
+		if strings.Index(annotationSpec, "=") != -1 {
+			parts := strings.SplitN(annotationSpec, "=", 2)
+			if len(parts) != 2 || len(parts[1]) == 0 {
+				return nil, nil, fmt.Errorf("invalid annotation spec: %v", annotationSpec)
+			}
+			annotations[parts[0]] = parts[1]
+		} else if strings.HasSuffix(annotationSpec, "-") {
+			remove = append(remove, annotationSpec[:len(annotationSpec)-1])
+		} else {
+			return nil, nil, fmt.Errorf("unknown annotation spec: %v", annotationSpec)
+		}
+	}
+	for _, removeAnnotation := range remove {
+		if _, found := annotations[removeAnnotation]; found {
+			return nil, nil, fmt.Errorf("can not both modify and remove an annotation in the same command")
+		}
+	}
+	return
+}
+
+func annotationFunc(obj runtime.Object, overwrite bool, resourceVersion string, annotations map[string]string, remove []string) (runtime.Object, error) {
+	meta, err := api.ObjectMetaFor(obj)
+	if err != nil {
+		return nil, err
+	}
+	if !overwrite {
+		if err := validateNoAnnotationOverwrites(meta, annotations); err != nil {
+			return nil, err
+		}
+	}
+
+	if meta.Annotations == nil {
+		meta.Annotations = make(map[string]string)
+	}
+
+	for key, value := range annotations {
+		meta.Annotations[key] = value
+	}
+	for _, annotation := range remove {
+		delete(meta.Annotations, annotation)
+	}
+
+	if len(resourceVersion) != 0 {
+		meta.ResourceVersion = resourceVersion
+	}
+	return obj, nil
+}
+
+// RunAnnotation executes kubectl annotation command
+func RunAnnotation(f *cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []string) error {
```

Comment:
We're trying make the command line tools more re-useable and composable (see https://github.com/GoogleCloudPlatform/kubernetes/issues/7311).  There a couple ways to do it, but one we've had good luck with in OpenShift is to create an `AnnotateOptions` object.  On that object, have the data required to perform the operation (`RESTClient`, `NewAnnotations`, `RemoveNames`, etc).  A `Complete` method adapts from the command line args and factory to the data required.  A `Validate` method checks to the `AnnotateOptions` to see if there is sufficient information run the command. And `RunAnnotation()` (no args), does the work.

Here's an example of one: https://github.com/openshift/origin/blob/master/pkg/cmd/cli/cmd/deploy.go#L81